rainier-core/src/main/scala/com/stripe/rainier/sampler/Ehmc.scala

```diff
@@ -0,0 +1,50 @@
+package com.stripe.rainier.sampler
+
+import scala.collection.mutable.ListBuffer
+import com.stripe.rainier.compute._
+import com.stripe.rainier.core._
+
+/**
+  * Empirical HMC - automated tuning of step size and number of leapfrog steps
+  * @param l0 the initial number of leapfrog steps to use during the dual averaging phase
+  * @param k the number of iterations to use when determining the
+  * empirical distribution of the total number of leapfrog steps until a u-turn
+  */
+final case class EHMC(l0: Int, k: Int = 2000) extends Sampler {
+  def sample(density: DensityFunction,
+             warmupIterations: Int,
+             iterations: Int,
+             keepEvery: Int)(implicit rng: RNG): List[Array[Double]] = {
+
+    val lf = LeapFrog(density)
+    val params = lf.initialize
+    val stepSize =
+      DualAvg.findStepSize(lf, params, 0.65, l0, warmupIterations)
+
+    val empiricalL: Vector[Int] = lf.empiricalLongestSteps(l0, k, stepSize)
```

Comment:
`findStepSize` acts as a warmup and advances the `params` vector. Two related questions:
* Do you want to be starting the empiricalLongestSteps trajectories from the vector that findStepSize ended with?
* Do you want to be using the empiricalLongestSteps work as additional warmup and continue the actual samples from the point it ends up at, or do you want to throw them away?

Either way you probably want to be passing in `params` here.